### PR TITLE
Transform form groups

### PR DESCRIPTION
**Pull request by GDeFranceschi**
In this PR, rebased on `transform-form`, I added methods to relate forms and groups of isometries preserved by the forms. The relevant novelties are in the file `src/Groups/matrices/forms.jl`, in the last two bunch of methods.
In the first bunch, there are functions to get the forms preserved by a group of matrices.
In the second bunch, the functions to compute the group of isometries of a given sesquilinear or quadratic form.

About the first one:
three approaches are shown. Which is the best one? We can also decide to keep more than one (e.g. sometimes one needs just any form preserved by the group, sometimes he may need all the forms).
Note that the first approach finds forms with every possible matrix, not just symmetric / hermitian / skew-symmetric, but I want to write a function to restrict to one of these cases.

About the second one:
this works only with sesquilinear or quadratic forms, so forms with Gram matrix being symmetric / hermitian / skew-symmetric. It works also with degenerate matrices.

Question: since the computation involved in these methods could be quite expensive, could it make sense to store the value of the preserved forms (or one of them) into a supplementary field of a matrix group (using `@declare_other`), analogously to what we did with the order?

### CODE REVIEW

File: src/Groups/matrices/forms.jl [hunk 1]
```diff
@@ -467,3 +468,311 @@ function issingular(f::SesquilinearForm{T}) where T
...
+#
+# From group to form #TODO: there are different approaches. Which is the best?
+#
+########################################################################
+
+# TODO 1st approach: brute force calculation
+# Algorithm furnished by Thomas Breuer, Aachen University
+# extended to quadratic by Giovanni De Franceschi, TU Kaiserslautern
+# WARNING: huge linear system !!
+
```
Comment (simonbrandhorst) [inline src/Groups/matrices/forms.jl]: Is this restricted to finite fields? To me it looks like the linear algebra approach could work just fine over any field or ZZ.
For instance it seems that magma can do it.
https://magma.maths.usyd.edu.au/magma/handbook/text/704

**Comment by fieker** [comment]
On Fri, Mar 19, 2021 at 11:56:14AM -0700, simonbrandhorst wrote:
> @simonbrandhorst commented on this pull request.
> 
> 
> 
> > @@ -467,3 +468,311 @@ function issingular(f::SesquilinearForm{T}) where T
>     f.descr != :quadratic && throw(ArgumentError("The form is not quadratic"))
>     return GAP.Globals.IsSingularForm(f.X)
>  end
> +
> +
> +
> +
> +########################################################################
> +#
> +# From group to form #TODO: there are different approaches. Which is the best?
> +#
> +########################################################################
> +
> +# TODO 1st approach: brute force calculation
> +# Algorithm furnished by Thomas Breuer, Aachen University
> +# extended to quadratic by Giovanni De Franceschi, TU Kaiserslautern
> +# WARNING: huge linear system !!
> +
> 
> Is this restricted to finite fields? To me it looks like the linear algebra approach could work just fine over any field or ZZ.
> For instance it seems that magma can do it.
> https://magma.maths.usyd.edu.au/magma/handbook/text/704
> 
My recollection, about Magma and such
 - lin alg would be best done over fin. field, then CRT to move to char
   0
 - at least some of this is either some form of Reynolds operator (in
   char 0 can be done using approximations, efficiently even for large
   groups, and even partially already there)  
   or MeatAxe

What exactly are you interested in? Examples of groups?

**Comment by fingolfin** [comment]
Please rebase now that the transform forms PR has been merged.

File: src/Groups/matrices/forms.jl [hunk 2]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+"""
+function invariant_quadratic_forms(G::MatrixGroup{S,T}) where {S,T}
+   F = base_ring(G)
+   n = degree(G)
+   M = T[]
+   for mat in gens(G)
+      MM = zero_matrix(F,div(n*(n+1),2),div(n*(n+1),2))
+      for i in 1:n, j in i:n
+      for p in 1:n, q in p:n
+         MM[div((2*n-i+2)*(i-1),2)+1, div((2*n-p+2)*(p-1),2)+q-p+1] = mat[i,p]*mat[i,q]
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: Hmm, `j` does not seem to be used here, am I missing something?

Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: You're right. It was just a leftover from a previous version. I'll remove it.

File: src/Groups/matrices/forms.jl [hunk 3]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
    f.descr != :quadratic && throw(ArgumentError("The form is not quadratic"))
    return GAP.Globals.IsSingularForm(f.X)
 end
+
+
+
+
+########################################################################
+#
+# From group to form #TODO: there are different approaches. Which is the best?
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: Did you run any benchmarks to compare these approaches with each other? Or do they all have the same performance ands this a question about "style"?

File: src/Groups/matrices/forms.jl [hunk 4]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+   for mat in gens(G)
+      MM = zero_matrix(F,div(n*(n+1),2),div(n*(n+1),2))
+      for i in 1:n
+      for p in 1:n, q in p:n
+         MM[div((2*n-i+2)*(i-1),2)+1, div((2*n-p+2)*(p-1),2)+q-p+1] = mat[i,p]*mat[i,q]
+      end
+      for j in i+1:n, p in 1:n, q in p:n
+         MM[div((2*n-i+2)*(i-1),2)+j-i+1, div((2*n-p+2)*(p-1),2)+q-p+1] = mat[i,p]*mat[j,q]+mat[j,p]*mat[i,q]
+      end
+      end
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: So perhaps reformat and merge these, and give names to some of the intermediate value? Something like this (please double check I didn't mess it up)
```suggestion
      for i in 1:n
         row = div((2*n-i+2)*(i-1),2)+1
         for p in 1:n, q in p:n
            col = div((2*n-p+2)*(p-1),2)+q-p+1
            MM[row, col] = mat[i,p]*mat[i,q]
            for j in i+1:n
               MM[row+j-i, col] = mat[i,p]*mat[j,q]+mat[j,p]*mat[i,q]
            end
         end
      end
```

File: src/Groups/matrices/forms.jl [hunk 5]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+      for i in 1:div(n*(n+1),2) MM[i,i]-=1 end
+      push!(M,MM)
+   end
+
+   r,K = nullspace(block_matrix(length(M),1,M))
+   M = T[]
+   for i in 1:r
+      push!(M,upper_triangular_matrix([K[j,i] for j in 1:div(n*(n+1),2)]))
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: I think this might also work (please check)
```suggestion
      push!(M,upper_triangular_matrix(K[1:div(n*(n+1),2),i]))
```

File: src/Groups/matrices/forms.jl [hunk 6]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+end
+
+"""
+    invariant_alternating_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of alternating forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_alternating_forms(G::MatrixGroup{S,T}) where {S,T}
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: Please add tests for this function.

Comment (fingolfin) [inline src/Groups/matrices/forms.jl]:
```suggestion
function invariant_alternating_forms(G::MatrixGroup)
```

File: src/Groups/matrices/forms.jl [hunk 7]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+#TODO: do we want to keep these?
+#TODO: hermitian forms require resolution of semilinear systems
+"""
+    invariant_symmetric_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of symmetric forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_symmetric_forms(G::MatrixGroup{S,T}) where {S,T}
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: Please add tests for this function.

File: src/Groups/matrices/forms.jl [hunk 8]
```diff
@@ -467,3 +468,399 @@ function issingular(f::SesquilinearForm{T}) where T
...
+end
+
+"""
+    invariant_quadratic_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of quadratic forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_quadratic_forms(G::MatrixGroup{S,T}) where {S,T}
```
Comment (fingolfin) [inline src/Groups/matrices/forms.jl]: Please add tests for this function.

**Review by fingolfin — CHANGES_REQUESTED** [review]
Lots and lots of code in here has no tests... Do you plan to change that? Or is that because you first want to know which of it to keep?

File: src/Groups/matrices/forms.jl (hunk 4 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: Done. It works.

File: src/Groups/matrices/forms.jl (hunk 5 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: It does not work because `K[1:n,2]` is a `MatElem` and not a vector. Unless I modify the function `upper_triangular_matrix` in order to make it work also for matrices.

File: src/Groups/matrices/forms.jl (hunk 6 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: I used `T` in the third line of the function, where I define `M = T[]`. Do I remove the parameters and replace the third line by `M = mat_elem_type(G)[]` ?
(the same correction will apply to the other analogous functions)

File: src/Groups/matrices/forms.jl (hunk 3 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: I took some experiments. The second approach (using GAP MeatAxe) is the fastest and the cheapest, but it works only for irreducible `G`-modules (so, it is useful to get quickly a form of a group we already know the nature).
The first approach uses more memory than the third one, but a less amount of time.
Anyway, the approaches have different aims. The first one produces all the forms of a certain type preserved by `G` (i.e. a basis for the vector space of all the forms preserved by `G`). The second one returns just one form preserved by `G`. The third one returns some of them, but not all of them.
If I'd choose, I would eliminate at least the third one (using GAP package "forms"), since it is not completely clear to me how it works, and keep the first approach (useful for a group recognition) and the second one (quicker). But it is just my opinion. Maybe you or @ThomasBreuer  have different ideas.

File: src/Groups/matrices/matrix_manipulation.jl [hunk 9]
```diff
@@ -122,6 +124,50 @@ function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <
...
+
+Return the upper triangular matrix whose non-zero entries (ordered lexicographically) are the elements of the sequence `L`.
+
+An error is returned whenever the length of `L` has not the form `n*(n+1)/2` for some integer `n`.
+"""
+function upper_triangular_matrix(L::AbstractVector{T}) where T <: RingElem
+   d = 1                # dimension of the matrix
+   while div(d*(d+1),2)< length(L)
+      d+=1
+   end
```
Comment (fingolfin) [inline src/Groups/matrices/matrix_manipulation.jl]: `d*(d+1)/2 == length(L)` => `(2d+1)^2 == 4d^2 + 4d + 1 == 8*length(L)+1` => `d = (sqrt(8*length(L)+1)-1)/2`
So modulo mistakes I made...
```suggestion
   d = (sqrt(8*length(L)+1)-1)/2
```

File: src/Groups/matrices/matrix_manipulation.jl [hunk 10]
```diff
@@ -122,6 +124,50 @@ function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <
...
 end
 
+"""
+    upper_triangular_matrix(L::AbstractVector{T}) where T <: RingElem
+
+Return the upper triangular matrix whose non-zero entries (ordered lexicographically) are the elements of the sequence `L`.
+
+An error is returned whenever the length of `L` has not the form `n*(n+1)/2` for some integer `n`.
+"""
+function upper_triangular_matrix(L::AbstractVector{T}) where T <: RingElem
```
Comment (fingolfin) [inline src/Groups/matrices/matrix_manipulation.jl]: The description refers to a sequence, and indeed, you could easily generalize this to work with arbitrary sequences of RingElem, including e.g. `fmpz_mat` and other Oscar vector-as-matrix types...
```suggestion
function upper_triangular_matrix(L)
   T = eltype(L)
   @assert T <: RingElem "L must be a collection of ring elements"
```

File: src/Groups/matrices/matrix_manipulation.jl [hunk 11]
```diff
@@ -122,6 +124,50 @@ function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <
...
+   d = 1                # dimension of the matrix
+   while div(d*(d+1),2)< length(L)
+      d+=1
+   end
+   length(L)==div(d*(d+1),2) || throw(ArgumentError("Input vector of invalid length"))
+   x = zero_matrix(parent(L[1]),d,d)
```
Comment (fingolfin) [inline src/Groups/matrices/matrix_manipulation.jl]:
```suggestion
   R = parent(first(L))
   x = zero_matrix(R,d,d)
```

File: src/Groups/matrices/matrix_manipulation.jl [hunk 12]
```diff
@@ -122,6 +124,50 @@ function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <
    return X
 end
 
+"""
+    upper_triangular_matrix(L::AbstractVector{T}) where T <: RingElem
+
+Return the upper triangular matrix whose non-zero entries (ordered lexicographically) are the elements of the sequence `L`.
```
Comment (fingolfin) [inline src/Groups/matrices/matrix_manipulation.jl]: This text doesn't quite make sense if L contains zeroes.
```suggestion
Return the upper triangular matrix whose entries on and above the diagonal are the elements of the sequence `L`. 
```

File: src/Groups/matrices/form_group.jl [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+   end
+
+   r,K = nullspace(block_matrix(length(M),1,M))
+
+   L = T[]
+   for j in 1:r
+      B = upper_triangular_matrix([K[i,j] for i in 1:div(n*(n+1),2)])
+      for i in 1:n, j in 1:i-1
+         B[i,j]=B[j,i]
+      end
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]: So... `invariant_quadratic_forms` and `invariant_symmetric_forms` use very different code style, but... aren't they in the end doing almost the exact same thing other than the above three lines?

If so, you could simplify `invariant_symmetric_forms` substantially by just calling `invariant_quadratic_forms` and then transforming the matrices in it with the above three lines...?

(Of course you can decide which version of the actual computation code to keep)

File: src/Groups/matrices/form_group.jl [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+
+   r,K = nullspace(block_matrix(length(M),1,M))
+   
+   return [matrix(F,n,n,[K[i,j] for i in 1:n^2]) for j in 1:r]
+end
+
+"""
+    invariant_sesquilinear_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of sesquilinear non-bilinear forms preserved by the group `G`. It works only if `base_ring(G)` has even degree.
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]: ... what's the degree of an arbitrary ring? I think you mean to say that `base_ring(G)` must be a finite field with an involution, right?

File: src/Groups/matrices/form_group.jl [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_sesquilinear_forms(G::MatrixGroup{S,T}) where {S,T}
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:63]: ... so I guess this would detect it?
```suggestion
function invariant_sesquilinear_forms(G::MatrixGroup{S,T}) where {S<:FinFieldElem,T}
```

Alternatively, it might be better to add a second argument in which the user is supposed to provide a "field automorphism" (which could simply be a function acting as such)... and as soon as you have *that*, you can reimplement `invariant_bilinear_forms` as
```
invariant_bilinear_forms(G::MatrixGroup) = invariant_sesquilinear_forms(G, identity)
```

File: src/Groups/matrices/form_group.jl [hunk 16]
```diff
@@ -0,0 +1,629 @@
...
+end
+
+"""
+    invariant_hermitian_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of hermitian forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
+"""
+function invariant_hermitian_forms(G::MatrixGroup{S,T}) where {S,T}
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:195]: Restrict to finite field elements?

File: src/Groups/matrices/form_group.jl [hunk 17]
```diff
@@ -0,0 +1,629 @@
...
+               pos_c +=1
+            end
+            pos_r +=1
+            pos_c =1
+         end
+ 
+         for i in 1:n^2
+            MM[i,i] -=1
+         end
+         push!(M,MM)
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:378]: This function is ~180 lines long, with no explanation as to what goes on, much less a source :/.

I think I could figure out the details to verify this, but then I dabbled in forms quite a bit in the past. A casual reader of this code will be completely lost here. So it would be great if at the start you could describe the Ansatz being used, in a code comment. The same for the other ,simpler functions (which find quadratic/symmetric/alternating/... forms)

File: src/Groups/matrices/form_group.jl [hunk 18]
```diff
@@ -0,0 +1,629 @@
...
+      if isodd(q)
+         for i in 1:n, j in i+1:n
+            sol[i,j] = e(K[pos,r])+w*e(K[pos+N,r])       # sol[i,j] = a_ij + w*b_ij
+            sol[j,i] = e(K[pos,r])-w*e(K[pos+N,r])
+            pos +=1
+         end
+      else
+         for i in 1:n, j in i+1:n
+            sol[i,j] = e(K[pos,r])+w*e(K[pos+N,r])       # sol[i,j] = a_ij + w*b_ij
+            sol[j,i] = e(K[pos,r])+e(K[pos+N,r])+w*e(K[pos+N,r])
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]: So... the same as this?
```suggestion
            sol[j,i] = e(K[pos,r])+(1+w)*e(K[pos+N,r])
```

I am lost, why is this code right?

File: src/Groups/matrices/form_group.jl [hunk 19]
```diff
@@ -0,0 +1,629 @@
...
+end
+
+"""
+    function invariant_quadratic_form(G::MatrixGroup)
+
+Return an invariant bilinear form for the group `G`. It works only if the module induced by the action of `G` is absolutely irreducible.
+!!! warning "Note:"
+    At the moment, the output is returned of type `mat_elem_type(G)`.
+"""
+function invariant_quadratic_form(G::MatrixGroup)
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:450]: Please add tests for this, and also take a look at the codecov comments visible on https://github.com/oscar-system/Oscar.jl/pull/368/files to discover more untested coode.

File: src/Groups/matrices/form_group.jl [hunk 20]
```diff
@@ -0,0 +1,629 @@
...
+      return m
+   end
+end
+
+
+# TODO 3rd approach: using GAP package "forms"
+"""
+    preserved_quadratic_forms(G::MatrixGroup)
+
+Uses random methods to find all of the quadratic forms preserved by `G` up to a scalar (i.e. such that `G` is a group of similarities for the forms). Since the procedure relies on a pseudo-random generator, the user may need to execute the operation more than once to find all invariant quadratic forms.
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]: Please do not use such long lines. Let's wrap them earlier; Julia manual recommends 92 chars as limit, so we could use e.g. that.

Please look at the other docstrings and also rewrap them if needed.

File: src/Groups/matrices/matrix_manipulation.jl (hunk 12 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/matrix_manipulation.jl]: Do I remove the `(ordered lexicographically)` ?

File: src/Groups/matrices/form_group.jl (hunk 13 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl]: Yes, we can do it. Unless we decide to consider symmetric forms in even characteristic (that are usually not taken in consideration in the study of classical groups). Do I simplify `invariant_symmetric_forms` by making it to call `invariant_quadratic_forms`?

File: src/Groups/matrices/form_group.jl (hunk 14 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl]: This kind of forms can be defined over any field with an involutory automorphism (e.g. the complex field). But these methods work only on finite fields. I will put a restriction on the type.

File: src/Groups/matrices/form_group.jl (hunk 18 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl]: The matrix `sol` must be hermitian, so `sol[j,i]` is equal to `sol[i,j]^q` for every `i,j`. Now, `(a_ij+w*b_ij)^q = a_ij+w^q*b_ij`. We built `w` in order to have `w^q=w+1`. It follows that if `sol[i,j]=a_ij+w*b_ij`, then `sol[j,i] = a_ij+b_ij+w*b_ij`.

File: src/Groups/matrices/form_group.jl (hunk 17 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl:378]: I've added some comments.

File: src/Groups/matrices/matrix_manipulation.jl (hunk 12 above)
Comment (fingolfin) [reply src/Groups/matrices/matrix_manipulation.jl]: Yes; and also the "non-zero" is misleading, as L could contain zeros. So you could just take my complete suggestion.

File: src/Groups/matrices/form_group.jl (hunk 13 above)
Comment (fingolfin) [reply src/Groups/matrices/form_group.jl]: Yes, I would suggest that you do that

File: src/Groups/matrices/form_group.jl (hunk 14 above)
Comment (fingolfin) [reply src/Groups/matrices/form_group.jl]: I know -- my questions are kinda rhetorical. Let me rephrase then: "This text is confusing because it talks about the degree of a ring, which is AFAIK not a general concept. I think what you really want to say here is that this method expects the base ring to be a finite field which has even degree over its prime field. Then you should say that, and also restrict the signature accordingly."

Well, or else, make the function generic, as I then suggested a few lines later...

File: src/Groups/matrices/form_group.jl (hunk 18 above)
Comment (fingolfin) [reply src/Groups/matrices/form_group.jl]: These things (and much more) should be explained in some code comments

File: src/Groups/matrices/form_group.jl [hunk 21]
```diff
@@ -0,0 +1,654 @@
...
+   
+   return [matrix(F,n,n,[K[i,j] for i in 1:n^2]) for j in 1:r]
+end
+
+"""
+    invariant_quadratic_forms(G::MatrixGroup)
+
+Return a generating set for the vector spaces of quadratic forms preserved by the group `G`.
+!!! warning "Note:"
+    At the moment, elements of the generating set are returned of type `mat_elem_type(G)`.
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:89]: By the way, I don't quite understand these notes: When you say "At the moment", it makes me think "... so there is a plan to do something else", and I wonder "but what"... Care to elaborate what you have in mind there?

File: src/Groups/matrices/form_group.jl (hunk 13 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl]: I modified it. But I have to add that it works properly only in odd characteristic.

File: src/Groups/matrices/form_group.jl (hunk 16 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl:195]: I put an assertion two lines below.

File: src/Groups/matrices/form_group.jl (hunk 21 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl:89]: One may expect to have an output of type `SesquilinearForm` instead of `MatElem`. I don't know if in future this can change, i.e. one will change these functions in order to have an output of type `SesquilinearForm`, that's why I wrote "At the moment". For now, I can just write that the output has type `MatElem`.

File: src/Groups/matrices/forms.jl (hunk 3 above)
Comment (fingolfin) [reply src/Groups/matrices/forms.jl]: Can you explain which benchmarks you performed? Like, which inputs groups did you use (which degree, over which fields?), as depending on that, the outcome of the benchmarks might vary quite a bit...

The approach from the `forms` package is (AFAIK) motivated by constructive group recognition, where solving the full linear system of equations can be too slow for big input groups; while finding *some* invariant forms heuristically usually is enough, and at least in GAP is quite a bit faster, at least if the degree of the matrix group is sufficiently high (I don't recall, but let's say 100x100 and not just 5x5 :-) ). Of course the fact that one of your functions is implemented in Julia with Nemo and the other calls into GAP code also can have a major impact on relative speeds: even if the "forms" algorithm really is asymptotically faster, this may be masked by the Julia code having lower "constants" in their actual performance, and so may be faster for small(er) examples... 

That's why it's important to know what exactly you tested.

Comment (GDeFranceschi) [reply src/Groups/matrices/forms.jl]: I took tests in dimension up to 10; never tried in dimension 100.
I tested the three approaches on `GO(1,6,9)`, `GO(9,3)`, `GU(6,5)`, `GO(-1,10,4)`, `Sp(8,5)` and `GU(6,4)`. Should I try some higher dimension, or higher field size?

File: src/Groups/matrices/form_group.jl [hunk 22]
```diff
@@ -0,0 +1,622 @@
...
+         row = div((2*n-i+2)*(i-1),2)+1
+         for p in 1:n, q in p:n
+            col = div((2*n-p+2)*(p-1),2)+q-p+1
+            MM[row, col] = mat[i,p]*mat[i,q]
+            for j in i+1:n
+               MM[row+j-i, col] = mat[i,p]*mat[j,q]+mat[j,p]*mat[i,q]
+            end
+         end
+      end
+      for i in 1:div(n*(n+1),2) MM[i,i]-=1 end
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]: In `invariant_alternating_forms` you do the following instead. It looks simpler, but OTOH the code here may be more efficient. I don't care which one you use, I just find it strange to see the same task done in two different ways for no apparent reason... perhaps there is one I am simply missing?
```suggestion
      MM -= one(MM)
```

File: src/Groups/matrices/form_group.jl [hunk 23]
```diff
@@ -0,0 +1,622 @@
...
+Return an invariant bilinear form for the group `G`.
+It works only if the module induced by the action of `G` is absolutely irreducible.
+!!! warning "Note:"
+    At the moment, the output is returned of type `mat_elem_type(G)`.
+"""
+function invariant_quadratic_form(G::MatrixGroup)
+   if iseven(characteristic(base_ring(G)))
+      V = GAP.Globals.GModuleByMats(GAP.Globals.GeneratorsOfGroup(G.X), G.mat_iso.fr.codomain)
+      B = GAP.Globals.MTX.InvariantQuadraticForm(V)
+      return _upper_triangular_version(G.mat_iso(B))
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:454]: Please add a test for these lines.

File: src/Groups/matrices/form_group.jl [hunk 24]
```diff
@@ -0,0 +1,622 @@
...
+the user may need to execute the operation more than once to find all invariant quadratic forms.
+"""
+function preserved_quadratic_forms(G::MatrixGroup{S,T}) where {S,T}
+   L = GAP.Globals.PreservedQuadraticForms(G.X)
+   R = SesquilinearForm{S}[]
+   for f_gap in L
+      f = quadratic_form(G.mat_iso(GAP.Globals.GramMatrix(f_gap)))
+      f.X = f_gap
+      f.mat_iso = G.mat_iso
+      push!(R,f)
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:482]: Please add a test for these lines.

File: src/Groups/matrices/form_group.jl [hunk 25]
```diff
@@ -0,0 +1,622 @@
...
+function preserved_sesquilinear_forms(G::MatrixGroup{S,T}) where {S,T}
+   L = GAP.Globals.PreservedSesquilinearForms(G.X)
+   R = SesquilinearForm{S}[]
+   for f_gap in L
+      if GAP.Globals.IsHermitianForm(f_gap)
+         f = hermitian_form(G.mat_iso(GAP.Globals.GramMatrix(f_gap)))
+      elseif GAP.Globals.IsSymmetricForm(f_gap)
+         f = symmetric_form(G.mat_iso(GAP.Globals.GramMatrix(f_gap)))
+      elseif GAP.Globals.IsAlternatingForm(f_gap)
+         f = alternating_form(G.mat_iso(GAP.Globals.GramMatrix(f_gap)))
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl:504]: Please add a test for these lines.

File: src/Groups/matrices/form_group.jl [hunk 26]
```diff
@@ -0,0 +1,622 @@
...
+      fn = SesquilinearForm(submatrix(C,1,1,r,r),f.descr)
+   else
+      fn = f
+   end
+
+   e=0
+   if (fn.descr==:quadratic || fn.descr==:symmetric) && iseven(r)
+      if witt_index(fn)== div(r,2) e=1
+      else e=-1
+      end
```
Comment (fingolfin) [inline src/Groups/matrices/form_group.jl]:
```suggestion
      if witt_index(fn) == div(r,2)
         e = 1
      else
         e = -1
      end
```

File: src/Groups/matrices/matrix_manipulation.jl [hunk 27]
```diff
@@ -109,7 +113,8 @@ end
 """
     matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1})
 
-Return the matrix whose rows are the vectors in `A`. Of course, vectors in `A` must have the same length and the same base ring.
+Return the matrix whose rows are the vectors in `A`.
+Of course, vectors in `A` must have the same length and the same base ring.
```
Comment (fingolfin) [inline src/Groups/matrices/matrix_manipulation.jl]:
```suggestion
All vectors in `A` must have the same length and the same base ring.
```

Comment (fingolfin) [reply src/Groups/matrices/matrix_manipulation.jl]: That said, it doesn't really matter: one of the next steps after this PR is merged is to go over the code and replace all uses of `submatrix`, `insert_block`, `diagonal_join`, `block_matrix` by their Julia counterparts, and then remove them

**Review by fingolfin — COMMENTED** [review]
We are close to done, I think. Just a few more tests would be nice, unless you tell me that adding them is problematic?

File: src/Groups/matrices/form_group.jl (hunk 22 above)
Comment (GDeFranceschi) [reply src/Groups/matrices/form_group.jl]: The second one works perfectly.

File: test/Groups/forms.jl [hunk 28]
```diff
@@ -323,3 +323,280 @@ end
...
+   @testset for f in L
+       for g in gens(G)
+          @test g.elm*f*transpose(g.elm)==f
+       end
+   end
+   L = invariant_quadratic_forms(G)
+   @testset for m in L
+       Q = quadratic_form(m)
+       for g in gens(G)
+          @test Q^g==Q
```
Comment (fingolfin) [inline test/Groups/forms.jl:406]: This check confuses me. The condition on `g` preserving the quadratic form given by `Q` should be "for all row vectors `v`, we have `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` which in characteristic different from 2 is equivalent to `Q = g.Q.transpose(g)`. For a check that also works in char 2, see https://github.com/gap-system/gap/pull/4361/files

Now, if g is orthogonal with respect to the "standard" form, then `transpose(g) = inv(g)` and then the above check kinda makes sense; but not in general? Or am I missing something?

Anyway, I suggest we add (in a future PR!) a function like `preserves_form(form, matrix)` or `respects_form` or so, with methods for the various form types. Then the above line can become the following (and similar for the other tests)
```suggestion
          @test preserves_form(Q, g)
```

Comment (fingolfin) [reply test/Groups/forms.jl:406]: to clarify the last part of the above comment: I am imaging here that we have types `QuadracticForm`, `BilinearForm`, `SesquilinearForm`, which store a matrix, and possibly a ring involution; then one can have methods like
```julia
preserves_form(f::BilinearForm, g) = g.elm*f*transpose(g.elm)==f
```
I think this would be useful to have anyway; one could then also apply the forms quite nicely:
```julia
Q = QuadraticForm(...)
v = rand(vectorspace)
z = Q(v) # evaluate the form
f = SesquilinearForm(...)
w = rand(vectorspace)
z2 = f(v, w)
```

**Review by fingolfin — APPROVED** [review]
I'll just merge this now as-is, it's "good enough". Thank you, Giovanni!

File: test/Groups/forms.jl (hunk 28 above)
Comment (GDeFranceschi) [reply test/Groups/forms.jl:406]: > This check confuses me. The condition on `g` preserving the quadratic form given by `Q` should be "for all row vectors `v`, we have `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` which in characteristic different from 2 is equivalent to `Q = g.Q.transpose(g)`. For a check that also works in char 2, see https://github.com/gap-system/gap/pull/4361/files
> 
> Now, if g is orthogonal with respect to the "standard" form, then `transpose(g) = inv(g)` and then the above check kinda makes sense; but not in general? Or am I missing something?

I don't get exactly what you mean. Anyway, for quadratic forms, the condition  `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` for every vector `v` does not imply the identity between matrices `Q = g.Q.transpose(g)`; rather, it implies that `Q - g.Q.transpose(g)` is a skew-symmetric matrix, but not necessarily a zero matrix (`X` is skew-symmetric iff `X = -transpose(X)` and `X` has zeros on the diagonal). This is all implemented in the definition of the method `==` for forms, and it holds in either odd and even characteristic.
